atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/AsyncSessionManager.java

```diff
@@ -0,0 +1,314 @@
+                    + "with null session");
+        }
+    }
+
+    static class SimpleAddressTranslator implements AddressTranslator {
+
+        private final Map<String, InetSocketAddress> mapper;
+
+        SimpleAddressTranslator(CassandraKeyValueServiceConfig config) {
+            this.mapper = config.addressTranslation();
+        }
+
+        @Override
+        public void init(Cluster cluster) {
+        }
+
+        @Override
+        public InetSocketAddress translate(InetSocketAddress address) {
+            InetSocketAddress temp = mapper.getOrDefault(address.getHostString(), address);
+            return new InetSocketAddress(temp.getAddress(), temp.getPort());
+        }
+
+        @Override
+        public void close() {
+        }
+    }
+
+    // class fields and methods
+    private static final Logger log = LoggerFactory.getLogger(AsyncSessionManager.class);
+    private static final AtomicReference<AsyncSessionManager> FACTORY = new AtomicReference<>(null);
+
+    public static void initialize(TaggedMetricRegistry taggedMetricRegistry) {
+        Preconditions.checkState(
+                null == FACTORY.getAndUpdate(
+                        previous ->
+                                previous == null ? new AsyncSessionManager(taggedMetricRegistry) : previous
+                ),
+                "Already initialized");
+    }
+
+    public static AsyncSessionManager getOrInitializeAsyncSessionManager() {
+        return FACTORY.updateAndGet(previous -> {
+            if (previous == null) {
+                return new AsyncSessionManager();
+            } else {
+                return previous;
+            }
+        });
+    }
+
+    public static AsyncSessionManager getAsyncSessinManager() {
+        AsyncSessionManager factory = FACTORY.get();
+        Preconditions.checkState(factory != null, "AsyncSessionManager is not initialized");
+        return factory;
+    }
+
+    // instance fields and methods
+    private final TaggedMetricRegistry taggedMetricRegistry;
+    private final Cache<UniqueCassandraCluster, CassandraClusterSessionPair> clusters = Caffeine.newBuilder()
+            .weakValues()
+            .removalListener(
+                    (RemovalListener<UniqueCassandraCluster, CassandraClusterSessionPair>) (key, value, cause) -> {
+                        value.session().close();
+                        value.cluster().close();
+                    })
+            .build();
+
+    private final AtomicLong cassandraId = new AtomicLong();
+    // global for all sessions
+    private final AtomicLong sessionId = new AtomicLong();
+
+    private AsyncSessionManager(TaggedMetricRegistry taggedMetricRegistry) {
+        this.taggedMetricRegistry = taggedMetricRegistry;
+    }
+
+    private AsyncSessionManager() {
+        this.taggedMetricRegistry = null;
+    }
+
+    public AsyncClusterSession getAsyncSession(CassandraKeyValueServiceConfig config, Set<InetSocketAddress> servers) {
```

Comment:
After reading the PR, I'm a bit confused as to why `AsyncClusterSession` and `CassandraClusterSessionPair` are separate objects? if you go with the suggestion of `CassandraClusterSessionPair` -> `CassandraClusterSession` the lines are very blurred, thoughts on combining them?